src/learner.cc

```diff
@@ -479,21 +499,92 @@ class LearnerImpl : public Learner {
-        tparam_.tree_method == 5 || name_gbm_ == "gblinear") {
+  // Revise `tree_method` and `updater` parameters after seeing the training
+  // data matrix
+  inline void PerformTreeMethodHeuristic(DMatrix* p_train) {
+    if (name_gbm_ != "gbtree" || cfg_.count("updater") > 0) {
+      // 1. This method is not applicable for non-tree learners
+      // 2. This method is disabled when `updater` parameter is explicitly
+      //    set, since only experts are expected to do so.
       return;
     }
 
-    if (!p_train->SingleColBlock() && cfg_.count("updater") == 0) {
-      if (tparam_.tree_method == 2) {
-        LOG(CONSOLE) << "tree method is set to be 'exact',"
-                     << " but currently we are only able to proceed with "
-                        "approximate algorithm";
+    const TreeMethod current_tree_method = tparam_.tree_method;
+    if (rabit::IsDistributed()) {
+      /* Choose tree_method='approx' when distributed training is activated */
+      CHECK(tparam_.dsplit != 0)
+        << "Precondition violated; dsplit cannot be zero in distributed mode";
+      if (tparam_.dsplit == 1) {
+        LOG(FATAL) << "Column-wise data split is currently not supported";
       }
-      cfg_["updater"] = "grow_histmaker,prune";
+      switch (current_tree_method) {
+       case TreeMethod::kAuto:
+        LOG(CONSOLE) << "Tree method is automatically selected to be 'approx' "
+                        "for distributed training.";
+        break;
+       case TreeMethod::kApprox:
+        // things are okay, do nothing
+        break;
+       case TreeMethod::kExact:
+       case TreeMethod::kHist:
+        LOG(CONSOLE) << "Tree method was set to be '"
+                     << (current_tree_method == TreeMethod::kExact ?
+                        "exact" : "hist")
+                     << "', but only 'approx' is available for distributed "
+                        "training. The `tree_method` parameter is now being "
+                        "changed to 'approx'";
+        break;
+       case TreeMethod::kGPUExact:
+       case TreeMethod::kGPUHist:
+        LOG(FATAL) << "Distributed training is not available with GPU algoritms";
+        break;
+       default:
+        LOG(FATAL) << "Unknown tree_method ("
+                   << static_cast<int>(current_tree_method) << ") detected";
+      }
+      tparam_.tree_method = TreeMethod::kApprox;
+    } else if (!p_train->SingleColBlock()) {
+      /* Some tree methods are not available for external-memory DMatrix */
+      switch (current_tree_method) {
+       case TreeMethod::kAuto:
+        LOG(CONSOLE) << "Tree method is automatically set to 'approx' "
+                        "since external-memory data matrix is used.";
+        break;
+       case TreeMethod::kApprox:
+        // things are okay, do nothing
+        break;
+       case TreeMethod::kExact:
+        LOG(CONSOLE) << "Tree method was set to be 'exact', "
```

Comment:
Not a big issue, but here you use `LOG(CONSOLE)`, in the GPU case you use `LOG(FATAL)`, may I ask about the reason?

Reply:
@trivialfis I'm not sure if it's okay to silently change `tree_method` from `gpu_hist` to `approx` or `hist`. Users may find it jarring.

Reply:
My intuition is that it isn't too big of a leap to change from `exact` to `approx`, since both run on CPU. On the other hand, it is a big leap to go from `gpu_hist` to `approx`, since users would presumably expected to benefit from their nice GPUs.